### ISSUE TITLE
"Plugin's runtime wasn't added" in Webpack 2

### ISSUE THREAD

**Issue by 2WheelCoder** [NekR/offline-plugin#152]
`offline-plugin@4.5.3` is working for me with `webpack@2.1.0-beta.13`, but when I upgrade to any of the newer Webpack versions I get this error:

```
Error: offline-plugin: Plugin's runtime wasn't added to one of your bundle entries. See this https://goo.gl/YwewYp for details.
    at Compiler.<anonymous> (/Users/nickstevens/Repos/surge-ui/node_modules/offline-plugin/lib/index.js:279:20)
    at next (/Users/nickstevens/Repos/surge-ui/node_modules/webpack/node_modules/tapable/lib/Tapable.js:104:14)
    at callback (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/dist/index.js:71:17)
    at Object.finallyHandler (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/finally.js:39:23)
    at Object.tryCatcher (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/promise.js:510:31)
    at Promise._settlePromiseAt (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/promise.js:584:18)
    at Promise._settlePromises (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/promise.js:700:14)
    at Async._drainQueue (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/nickstevens/Repos/surge-ui/node_modules/copy-webpack-plugin/node_modules/bluebird/js/main/async.js:15:14)
    at tryOnImmediate (timers.js:543:15)
    at processImmediate [as _immediateCallback] (timers.js:523:5)
```

The runtime is included like this in my app's index file:
```
import OfflinePlugin from 'offline-plugin/runtime'
OfflinePlugin.install()
```

My `webpack.config.js` looks like this:
```js
const { resolve } = require('path')
const webpack = require('webpack')
const ExtractTextPlugin = require('extract-text-webpack-plugin')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const BrowserSyncPlugin = require('browser-sync-webpack-plugin')
const CopyWebpackPlugin = require('copy-webpack-plugin')
const Dotenv = require('dotenv-webpack')
const OfflinePlugin = require('offline-plugin')

require('dotenv').config({
  path: process.env.CONFIG
})

const HOST = process.env.HOST || 'localhost'
const PORT = process.env.PORT || 8080
const PROXY = `http://${HOST}:${PORT}`

const removeEmpty = array => array.filter(i => !!i)

module.exports = () => {
  return {
    entry: {
      app: process.env.NODE_ENV === 'dev' ?
        ['./index.js', 'webpack/hot/only-dev-server', 'webpack-dev-server/client?http://localhost:8080']
        : './index.js',
      vendor: ['react', 'redux', 'react-redux', 'axios', 'react-router', 'react-router-redux', 'immutable', 'react-dom']
    },
    output: {
      filename: process.env.NODE_ENV === 'dev' ? 'bundle.[name].js' : 'bundle.[name].[hash].js',
      path: resolve(__dirname, 'dist'),
      pathinfo: process.env.NODE_ENV === 'dev',
      publicPath: process.env.BASE_URL ? process.env.BASE_URL : '/'
    },
    resolve: {
      extensions: ['.jsx', '.scss', '.js', '.json']
    },
    context: resolve(__dirname, 'src'),
    devtool: process.env.NODE_ENV === 'dev' ? 'eval' : 'source-map',
    devServer: {
      historyApiFallback: true
    },
    bail: process.env.NODE_ENV !== 'dev',
    module: {
      loaders: [
        {
          test: /\.js$/,
          loader: 'react-hot!babel-loader!eslint',
          exclude: resolve(__dirname, "node_modules")
        },
        {
          test: /\.jsx$/,
          loader: 'react-hot!babel-loader?presets[]=react,presets[]=es2015-webpack,presets[]=stage-2,plugins[]=transform-class-properties!eslint',
          exclude: resolve(__dirname, "node_modules")
        },
        {
          test: /\.scss$/,
          loader: ExtractTextPlugin.extract('css-loader?sourceMap&modules&importLoaders=1&localIdentName=[name]__[local]___[hash:base64:5]!postcss-loader!sass-loader')
        },
        {
          test: /\.css$/,
          loader: ExtractTextPlugin.extract('css-loader')
        },
        {
          test: /\.icons\.(js|json)$/,
          loader: 'style-loader!css-loader!fontgen-loader'
        },
        {
          test: /\.json$/,
          loader: 'json-loader'
        },
        {
          test: /\.(jpe?g|png|gif|svg)$/i,
          loader: 'url?limit=10000!img?progressive=true'
        },
      ]
    },
    plugins: removeEmpty([
      new webpack.DefinePlugin({
        'process.env': {
          'NODE_ENV': JSON.stringify(process.env.NODE_ENV),
          'SURGE_HOME_URL': JSON.stringify(process.env.SURGE_HOME_URL),
          'SURGE_API_URL': JSON.stringify(process.env.SURGE_API_URL),
          'BASE_URL': JSON.stringify(process.env.BASE_URL)
        }
      }),
      new webpack.HotModuleReplacementPlugin(),
      new HtmlWebpackPlugin({
        template: './index.html'
      }),
      new BrowserSyncPlugin(
        {
          host: HOST,
          port: PORT,
          proxy: PROXY
        },
        {
          // prevent BrowserSync from reloading the page
          // and let Webpack Dev Server take care of this
          reload: false
        }
      ),
      new ExtractTextPlugin({
        filename: 'styles.[hash].css',
        allChunks: true
      }),
      new webpack.optimize.CommonsChunkPlugin({
        name: 'vendor',
      }),
      new CopyWebpackPlugin([
        {
          context: process.env.NODE_ENV === 'production' ?
            './browser/production/' :
            './browser/staging/',
          from: {
            glob: '**/*',
            dot: false
          }
        }
      ]),
      new OfflinePlugin({
        externals: ['https://fonts.googleapis.com/icon?family=Material+Icons'],
        ServiceWorker: {
          navigateFallbackURL: '/'
        }
      })
    ])
  }
}
```

Any ideas why this error is occurring even though I am definitely including the offline-plugin runtime? Thanks!

**Comment by NekR** [comment]
@2WheelCoder thanks for reporting this. Might be related to: https://github.com/NekR/offline-plugin/issues/148 somehow, but I'm not sure yet. I'll have to try with your config.

**Comment by NekR** [comment]
@2WheelCoder I don't see any problems with it in `webpack@2.2.0-rc.3`. Can you try in it? Also, how do you run your webpack build?

**Comment by NekR** [comment]
@2WheelCoder Also, are you running it with `dev-server`? `offline-plugin` doesn't work in dev-server. At least it's not supported yet and it worked before that was coincidence.

**Comment by maslianok** [comment]
@NekR 

Getting the same error.
Seems that this error appears only when using `import` instead of `System.import`.
But this feature was introduced only in webpack beta28 https://github.com/webpack/webpack/releases/tag/v2.1.0-beta.28

**Comment by NekR** [comment]
@maslianok did you delete your comment here? Can you confirm that you are also using `webpack-dev-server`?

**Comment by maslianok** [comment]
@NekR Sorry, I found that `import` doesn't work for me at all. So I think it's not a problem of `offline-plugin`.

No, I'm testing in production env.

**Comment by NekR** [comment]
@maslianok sure, no problem at all :-)

**Comment by NekR** [comment]
@2WheelCoder oh, I was able to repro this. Accidentally. I know what's going on and it's related to webpack 2 itself. In webpack 2, all loaders have to have full names, i.e. `-loader` isn't added automatically. In your project I found at least 2 such loaders where you don't add that prefix: `url` and `eslint`. If you will add it, then everything magically should start working. 

That's so weird behavior and I'll try to resolve it on webpack side. Thanks for reporting this!

**Comment by NekR** [comment]
Okay, I made it a bit different type of errors so it now more obvious what happens and doesn't exit whole program (since doesn't shallow webpack's `-loader` error), but that still webpack's issue and I hope it will be resolve prior webpack 2 goes to stable.

Commit: https://github.com/NekR/offline-plugin/commit/b84f4f1808b6cdd8b56c9f1ac8bd9d53200ed689

**Comment by 2WheelCoder** [comment]
Thanks @NekR! I thought I had updated the plugins that needed `-loader` appended, but it appears that all of them needed it and I had only updated a few. Everything seems to be working smoothly now with `webpack@2.2.0-rc.2`.

I am indeed using it with `webpack-dev-server`, but I hadn't had any issues before this. Seems to be working fine with `webpack-dev-server` now. This issue did not seem to be affected by it because I still received the error when running my staging and production builds without `webpack-dev-server`.
